src/ext/serializer.c

```diff
@@ -0,0 +1,130 @@
+#include <php.h>
+#include <stdio.h>
+#include "mpack/mpack.h"
+
+static void ddtrace_zval_to_writer(mpack_writer_t *writer, zval *trace);
+
+#if PHP_VERSION_ID < 70000
+static void ddtrace_hash_table_to_writer(mpack_writer_t *writer, HashTable *ht) /* {{{ */
+{
+    zval **tmp;
+    char *string_key;
+    uint str_len;
+    HashPosition iterator;
+    zend_ulong num_key;
+    int key_type;
+    bool first_time = true;
+
+    zend_hash_internal_pointer_reset_ex(ht, &iterator);
+    while (zend_hash_get_current_data_ex(ht, (void **)&tmp, &iterator) == SUCCESS) {
+        key_type = zend_hash_get_current_key_ex(ht, &string_key, &str_len, &num_key, 0, &iterator);
+        if (first_time == true) {
+            first_time = false;
+            if (key_type == HASH_KEY_IS_STRING) {
+                mpack_start_map(writer, zend_hash_num_elements(ht));
+            } else {
+                mpack_start_array(writer, zend_hash_num_elements(ht));
+            }
+        }
+        if (key_type == HASH_KEY_IS_STRING) {
+            mpack_write_cstr(writer, string_key);
+        }
+        ddtrace_zval_to_writer(writer, *tmp);
+        zend_hash_move_forward_ex(ht, &iterator);
+    }
+
+    if (key_type == HASH_KEY_IS_STRING) {
+        mpack_finish_map(writer);
+    } else {
+        mpack_finish_array(writer);
+    }
+}
+#else
+static void ddtrace_hash_table_to_writer(mpack_writer_t *writer, HashTable *ht) /* {{{ */
+{
+    zval *tmp;
+    zend_string *string_key;
+    zend_ulong num_key;
+    int is_assoc = -1;
+
+    ZEND_HASH_FOREACH_KEY_VAL_IND(ht, num_key, string_key, tmp) {
+        if (is_assoc == -1) {
+            is_assoc = string_key != NULL ? 1 : 0;
+            if (is_assoc == 1) {
+                mpack_start_map(writer, zend_hash_num_elements(ht));
+            } else {
+                mpack_start_array(writer, zend_hash_num_elements(ht));
+            }
+        }
+        if (is_assoc == 1) {
+            mpack_write_cstr(writer, ZSTR_VAL(string_key));
+        }
+        ddtrace_zval_to_writer(writer, tmp);
+    }
+    ZEND_HASH_FOREACH_END();
+
+    if (is_assoc) {
+        mpack_finish_map(writer);
+    } else {
+        mpack_finish_array(writer);
+    }
+}
+#endif
+
+static void ddtrace_zval_to_writer(mpack_writer_t *writer, zval *trace) /* {{{ */
```

Comment:
Variable named `trace` makes me think there might be something trace specific, while this is simply zval serializer.

`msgpack_write_zval` also would help with readability imho.

To avoid very long function names, I mostly tried to add `ddtrace_` prefix to non static functions, (with non static functions the probability of symbol name conflic exits, so the prefix needs to be there).